test/apply-policy.js

```diff
@@ -0,0 +1,278 @@
+'use strict'
+
+const { test } = require('tap')
+const Fastify = require('fastify')
+const mercurius = require('mercurius')
+const { GraphQLDirective } = require('graphql')
+const mercuriusAuth = require('..')
+
+test('basic - apply policy should provide authDirectiveAST', async (t) => {
+  t.plan(4)
+
+  const app = Fastify()
+  t.teardown(app.close.bind(app))
+
+  const schema = `
+  directive @auth on OBJECT | FIELD_DEFINITION
+
+  type Query {
+    add(x: Int, y: Int): Int @auth
+    subtract(x: Int, y: Int): Int
+  }
+`
+
+  const resolvers = {
+    Query: {
+      add: async (_, obj) => {
+        const { x, y } = obj
+        return x + y
+      },
+      subtract: async (_, obj) => {
+        const { x, y } = obj
+        return x - y
+      }
+    }
+  }
+
+  app.register(mercurius, {
+    schema,
+    resolvers
+  })
+  app.register(mercuriusAuth, {
+    authContext: (context) => {
+      return {
+        identity: context.reply.request.headers['x-user']
+      }
+    },
+    applyPolicy: async (authDirectiveAST, context, field) => {
+      t.equal(authDirectiveAST.kind, 'Directive')
+      t.equal(authDirectiveAST.name.value, 'auth')
+      t.same(authDirectiveAST.arguments, [])
+      return true
+    },
+    authDirective: new GraphQLDirective({ name: 'auth', locations: [] })
+  })
+
+  const query = `query {
+  four: add(x: 2, y: 2)
+  subtract(x: 3, y: 3)
+}`
+
+  const response = await app.inject({
+    method: 'POST',
+    headers: { 'content-type': 'application/json', 'X-User': 'ADMIN' },
+    url: '/graphql',
+    body: JSON.stringify({ query })
+  })
+
+  t.same(JSON.parse(response.body), {
+    data: {
+      four: 4,
+      subtract: 0
+    }
+  })
+})
+
+test('basic - apply policy should provide auth on context', async (t) => {
+  t.plan(2)
+
+  const app = Fastify()
+  t.teardown(app.close.bind(app))
+
+  const schema = `
+  directive @auth on OBJECT | FIELD_DEFINITION
+
+  type Query {
+    add(x: Int, y: Int): Int @auth
+    subtract(x: Int, y: Int): Int
+  }
+`
+
+  const resolvers = {
+    Query: {
+      add: async (_, obj) => {
+        const { x, y } = obj
+        return x + y
+      },
+      subtract: async (_, obj) => {
+        const { x, y } = obj
+        return x - y
+      }
+    }
+  }
+
+  app.register(mercurius, {
+    schema,
+    resolvers
+  })
+  app.register(mercuriusAuth, {
+    authContext: (context) => {
+      return {
+        identity: context.reply.request.headers['x-user']
+      }
+    },
+    applyPolicy: async (authDirectiveAST, context, field) => {
```

Comment:
have we got access to the current object/paths that lead to the current resolver? I think we need to express a concept like "this object is accessible only from owners of the parent org".

Reply:
good point! We have partial access to the current object through the `field` argument with is a `GraphQLField` type, I'm also building a cursor object as I've traverse the AST, so can also expose that in a sensible manner. What I'll do is change this to closely match with what is available in a normal resolver, e.g. `schema`, `info`, `context`, field etc.

I'm also adding support for the user to provide custom errors as I reckon this'll be needed